chips/nrf52840/src/ieee802154_radio.rs

```diff
@@ -767,128 +835,254 @@ impl<'a> Radio<'a> {
         buffer
     }
 
-    // TODO: Theres an additional step for 802154 rx/tx handling
+    fn crc_check(&self) -> Result<(), ErrorCode> {
+        if self.registers.crcstatus.is_set(Event::READY) {
+            Ok(())
+        } else {
+            Err(ErrorCode::FAIL)
+        }
+    }
+
+    // TODO: RECEIVING ACK FOR A SENT TX IS NOT IMPLEMENTED //
+    // As a general note for the interrupt handler, event registers must still be cleared even when
+    // hardware shortcuts are enabled.
     #[inline(never)]
     pub fn handle_interrupt(&self) {
         self.disable_all_interrupts();
 
-        if self.registers.event_ready.is_set(Event::READY) {
-            self.registers.event_ready.write(Event::READY::CLEAR);
-            self.registers.event_end.write(Event::READY::CLEAR);
-            if self.transmitting.get()
-                && self.registers.state.get() == nrf5x::constants::RADIO_STATE_RXIDLE
-            {
-                self.registers.task_ccastart.write(Task::ENABLE::SET);
-            } else {
-                self.registers.task_start.write(Task::ENABLE::SET);
+        let mut start_task = false;
+        let mut rx_init = false;
+
+        match self.state.get() {
+            // It should not be possible to receive an interrupt while the tracked radio state is OFF
+            RadioState::OFF => panic!("Received interrupt while off"),
+            RadioState::RX => {
+                ////////////////////////////////////////////////////////////////////////////
+                // NOTE: This state machine assumes that the READY_START shortcut is enabled
+                // at this point in time. If the READY_START shortcut is not enabled, the
+                // state machine/driver will likely exhibit undefined behavior.
+                ////////////////////////////////////////////////////////////////////////////
+
+                // Since READY_START shortcut enabled, always clear READY event
+                self.registers.event_ready.write(Event::READY::CLEAR);
+
+                // Completed receiving a packet, now determine if we need to send ACK
+                if self.registers.event_end.is_set(Event::READY) {
+                    self.registers.event_end.write(Event::READY::CLEAR);
+                    let result = self.crc_check();
+
+                    // Unwrap fail = Radio RX Buffer is missing (may be due to receive client not replacing in receive(...) method,
+                    // or some instance in  driver taking buffer without properly replacing).
+                    let rbuf = self.rx_buf.take().unwrap();
+
+                    let frame_len = rbuf[MIMIC_PSDU_OFFSET as usize] as usize - radio::MFR_SIZE;
```

Comment:
Should there be any sanity checking on this `frame_len`, or is that up to the user? For example, in the case that the CRC is bad, this could be incorrect or invalid (>127).

Reply:
I am unsure what is best here. On one hand, it makes sense to have the sanity check since the size is enforced by the 15.4 spec. The status of the CRC check is passed to the user/client so theoretically higher levels can handle this. I am curious as to your/others thoughts on if this should be checked here or handled by the client

Reply:
Recall that this is a hardware specific implementation of a general interface, not an abstraction layer, so it is not in the business of helping clients avoid shooting themselves in the foot (that may or may not be a job for middle layers).

This should reliably imementation the HIL, without unnecessarily panicking or otherwise breaking the rest of the system.

I'm making some assumptions here without fully being switched into to the specifics of the interface, but I think that means that if a client passed in an invalid frame (e.g. it's too long), that's the client's business.

Reply:
That all makes sense to me. The client is aware of whether the CRC is valid or not, and can make decisions based on that information (like not trusting the length field).